log/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+- `go.opentelemetry.io/otel/log`
+- `go.opentelemetry.io/otel/log/embedded`
+- `go.opentelemetry.io/otel/log/noop`
+
+Rejected alternative:
+
+- [Reuse slog](#reuse-slog)
+
+### LoggerProvider
+
+The [`LoggerProvider` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#loggerprovider)
+is defined as an interface:
+
+```go
+type LoggerProvider interface {
+	embedded.LoggerProvider
+	Logger(name string, options ...LoggerOption) Logger
+}
+```
+
+The specification may add new operations to `LoggerProvider`.
+The interface may have methods added without a package major version bump.
+The embedds `embedded.LoggerProvider` to help inform an API implementation
+author about this non-standard API evolution.
+This approach is already used in Trace API and Metrics API.
+
+#### LoggerProvider.Logger
+
+The `Logger` method implements the [`Get a Logger` operation](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#get-a-logger).
+
+Implementation requirements:
+
+- The [specification requires](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#concurrency-requirements)
+  the method to be safe to be called concurrently.
+
+- The method should use some default name if the passed name is empty
+  in order to meet the [specification's SDK requirement](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logger-creation)
+  to return a working logger when an invalid name is passed
+  as well as to resemble the behavior of getting tracers and meters.
+
+`Logger` can be extended by adding new `LoggerOption` options
+and adding new exported fields to the `LoggerConfig` struct.
+This design is already used in Trace API for getting tracers
+and in Metrics API for getting meters.
+
+Rejected alternative:
+
+- [Passing struct as parameter to LoggerProvider.Logger](#passing-struct-as-parameter-to-loggerproviderlogger).
+
+### Logger
+
+The [`Logger` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#logger)
+is defined as an interface:
+
+```go
+type Logger interface {
+	embedded.Logger
+	Emit(ctx context.Context, record Record)
+}
+```
+
+The specification may add new operations to `Logger`.
+The interface may have methods added without a package major version bump.
+The embedds `embedded.Logger` to help inform an API implementation
+author about this non-standard API evolution.
+This approach is already used in Trace API and Metrics API.
+
+### Logger.Emit
+
+The `Emit` method implements the [`Emit a LogRecord` operation](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord).
+
+Calls to `Emit` are supposed to be on the hot path.
+Therefore, in order to reduce the number of heap allocations,
+the [`LogRecord` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord),
+is defined as a `struct`:
+
+```go
+type Record struct {
+	Timestamp         time.Time
+	ObservedTimestamp time.Time
+	Severity          Severity
+	SeverityText      string
+	Body              string
```

Comment:
I know noticed that [the spec defines](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body) Body to be [`any`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#type-any) (not a `string`.)

[Java](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/logs/LogRecordBuilder.java) (stable logs): Defines Body as `string`.


[.NET](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Logs/LogRecordData.cs) (stable logs): Defines Body as `string`.

[C++](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Logs/LogRecordData.cs) (stable logs): Defines Body as attribute value and [a map as value is not supported](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/api/include/opentelemetry/common/attribute_value.h).

[Python](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/_logs/_internal/__init__.py) (experimental logs): Defines Body as `Any` (in Python sense which means it can be anything).

Reply:
I propose to change the `Body` type from `string` to `attribute.Value`.

Additionally, regarding having maps as values. 

I also propose create an issue in the specification (and add it to the project) to clarify if type is intended to be the same as attribute value type or something different. I would also reference the issue in the design doc for additional tracking. This is also related to both open-telemetry/opentelemetry-specification#376 and open-telemetry/opentelemetry-specification#622.

Reply:
CC @tigrannajaryan

Reply:
After reading the spec again I see that https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-body says:

> First-party Applications SHOULD use a string message. However, a structured body SHOULD be used to preserve the semantics of structured logs emitted by Third-party Applications. 

From my understanding of https://opentelemetry.io/docs/specs/otel/logs/#new-first-party-application-logs, the First-party Applications are the ones that use the Bridge API. Does it mean that Body should be defined as `string` in Bridge API (as done in Java and .NET)?

Reply:
https://github.com/pellared/opentelemetry-go/pull/5 shoes that having `Record.Body` as `any` ("object") would reduce the performance. Therefore, it makes more sense to define it as a `string` especially that I am not aware of any Go library that accepts `any` as a log record (it is always a text message with key-value attributes).

Reply:
Besides [Logs Data Model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body) defines Body to be [any](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#type-any). However, IMO the definition of `any` is wrong. It should be something that is used in given language to pass any type of data. e.g. in Java and .NET it is object , in Go any , in Python Any, in C++ probably std::any. The exporters would be responsible for marshaling the Body to expected output (e.g. text form, or OTLP protobuf).

Reply:
I documented why `Body` us defined as `string` and not `any`.

I plan to make more clarifications in [Logs Data Model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body).

Reply:
After sync with @jack-berg it occurs that using `attribute.Value` (like in C++) might be more appropriate. I think I will also explore this idea.

> The idea is that some log APIs allow recording of structured messages, and `AnyValue` allows that structure to be retained in the body, instead of having it flattened to a string.
> 
> Accepting object forces exporters to do to much. Its unbounded and serializing it would be very brittle. `AnyValue` has bounds - it has an enumeration of allowed types.
> 
> So serializing it much more simple, requiring just a switch on the type. `AnyValue` also has a helpful asString() method to serialize in a string representation. Exporters which don’t support structured values can use this to retain all the information in an encoded format.

Reply:
I do not think limiting `Body` to `string` is correct. It is not mandatory for applications to use string, it is a recommendation that may be violated with good justification (hence SHOULD and not MUST). The [experimental Events API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md) that is built on top of Log Bridge API will likely use non-string Body (this is still under a debate).

An alternate approach that may be worth considering is to have more than one variation of Emit method if you cannot make one Emit function performant enough for all use cases, each variation support a different set of parameters or different types for Body, etc. This is not prohibited by the spec. I don't like it but it is a possible option to consider.

`any` is a type defined in the spec. It matches exactly the definition of [AnyValue](https://github.com/open-telemetry/opentelemetry-proto/blob/c56093a326ec7d9c2436c778c50f656cb73b718c/opentelemetry/proto/common/v1/common.proto#L28) in OTLP. We made a choice of what types we want to be representable by `any`. This is part of a Stable spec doc, so breaking changes are no longer allowed.

How exactly you model `any` type in the language implementation is the particular SDK's design choice.

If you want to use Go's `any` type for it it is up to you, just be aware that you must limit the _output_ values to what is defined by the spec. This is not an exporter responsibility, it is the API implementation responsibility. The API implementation is not allowed to accept data types that are not defined in the spec's data model (for example you cannot accept time.Duration as a value because data model does not allow it).

Alternatively you can define AnyValue as a strongly-typed Go struct that can represent all valid types.

Reply:
>However, I am not sure how a bridge would now how to construct a attribute.Value from some abstract structured-body.

That is defined by the spec: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-type-mapping.md

Reply:
In java, the log bridge API has methods to set the body as a string or AnyValue. The AnyValue method is currently experimental and requires jumping through extra hoops to use. The string method is essentially syntactic sugar for setting the Body to an AnyValue string type. The string method will be the most commonly used, but its important to have the AnyValue method to preserve the structure of any logs recorded using structured logging APIs.

Reply:
> I do not think limiting `Body` to `string` is correct.

This is not clearly defined in the specification. See: https://cloud-native.slack.com/archives/C01N7PP1THC/p1704907537212259

Personally, I would add another field e.g. `BodyStructured attribute.Value` to `Record`. The user should set only `Body` or `BodyStructured`. But this can be added later as it is not a breaking change. Also it would be good to resolve open-telemetry/opentelemetry-specification#376 before that.

Reply:
I created https://github.com/open-telemetry/opentelemetry-specification/pull/3814

Reply:
I also created https://github.com/open-telemetry/opentelemetry-specification/pull/3827

Besides I think it would be bad to define `Body` as `attribute.Value`. I think it would be better and safer to have a separate type and not couple Body value to attribute representation.

Reply:
After today's Specification SIG meeting I plan to define `Body` as a type that would enable representing a structured log which will be different than `attribute.Value`.

Reply:
PR for the prototype https://github.com/pellared/opentelemetry-go/pull/7